erpnext/hr/doctype/expense_claim/expense_claim.py

```diff
@@ -36,22 +36,22 @@ def validate(self):
 			self.project = frappe.db.get_value("Task", self.task, "project")
 
 	def set_status(self):
-		self.status = {
+		status = {
 			"0": "Draft",
 			"1": "Submitted",
 			"2": "Cancelled"
 		}[cstr(self.docstatus or 0)]
 
 		paid_amount = flt(self.total_amount_reimbursed) + flt(self.total_advance_amount)
 		precision = self.precision("grand_total")
-		if (self.is_paid or (flt(self.total_sanctioned_amount) > 0
-			and flt(self.grand_total, precision) ==  flt(paid_amount, precision))) \
-			and self.docstatus == 1 and self.approval_status == 'Approved':
-				self.status = "Paid"
+		if (self.is_paid or (flt(self.total_sanctioned_amount) > 0 and self.docstatus == 1
+			and flt(self.grand_total, precision) == flt(paid_amount, precision))) and self.approval_status == 'Approved':
+			status = "Paid"
 		elif flt(self.total_sanctioned_amount) > 0 and self.docstatus == 1 and self.approval_status == 'Approved':
-			self.status = "Unpaid"
+			status = "Unpaid"
 		elif self.docstatus == 1 and self.approval_status == 'Rejected':
-			self.status = 'Rejected'
+			status = 'Rejected'
+		self.db_set("status", status)
```

Comment:
Since `set_status` is also called on validate when changes are not committed to the db, we can use a flag `update=False` in the method to understand whether to use `db_set` or not. Pass True to `set_status` when called from `on_submit`. Keep it false by default.